### PR TITLE
Migrate to Symfony Mailer

### LINKED ISSUE #29731: Migrate from Swiftmailer to Symfony Mailer
We currently use https://github.com/swiftmailer/swiftmailer which is stopping development at the end of November and advising to move to Symfony Mailer.

There are rectors for this but I did not manage to get them to work on our code as-is:
https://github.com/rectorphp/rector-symfony/blob/main/config/sets/swiftmailer/swiftmailer-to-symfony-mailer.php
https://github.com/rectorphp/rector-symfony/blob/main/src/Rector/MethodCall/SwiftSetBodyToHtmlPlainMethodCallRector.php

Switching from `\Swift_Message` to `\Symfony\Component\Mime\Email` is pretty straightforward and should not affect our public interfaces in \OCP\Mail.
The transport part may be less straightforward, but both sendmail and smtp are supported by symfony mailer.

### PR DESCRIPTION
**Pull request by come-nc**
Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

Fixes https://github.com/nextcloud/server/issues/29731

TODO:
- [x] Figure out if `mail_smtpsecure` should be dropped or re-implemented (see https://github.com/nextcloud/server/pull/30349#issuecomment-1268107105 and https://github.com/nextcloud/server/pull/30349#discussion_r1028074804 )
    - [x]  https://symfony.com/doc/current/mailer.html#tls-peer-verification
    - [x] https://github.com/symfony/symfony/issues/41997#issuecomment-874733474
- [x] Drop `mail_smtpauthtype`
- [x] Failed recipients are back 
- [x] Test that mail sending still works
- [x] Tested with gmx.net:578 (starttls) account
    - [x] Password reset email
    - [x] New account created email (Welcome aboard)
    - [x] Calendar events
    - [x] Activity emails
    - [x] File share by email
    - [x] announcement center
    - [x] Talk email
    - [x] Booking emails
    - [x] Remote wipe email - nothing in the logs,
    - [x] Email verification email
- [x] Tested with SMTP ssl/tls by (@come-nc )
    - [x] Password reset email
    - [x] Activity emails
    - [x] File share by email
- [x] Remove workaround code pointed by https://github.com/nextcloud/server/pull/30349#issuecomment-1268135892
    - [x] https://github.com/j-bernard/symfony/commit/f19d064e60cd4de3b57b923731ed011003db65c2 looks like IDN conversion works out of the box.
    - [x] validation package works with IDNs too
    - [x] Test
- [x] Testmail doesn't work as our `\OC\Mail\Message::setTo` function doesn't provide the right array structure for Symphony's send. WORKS
- [x] To get the behaviour identical to the old mailer, the wrapping in an `Address` object will be done at send. 
- [x] The email generation code needs a try - catch wrap now as there will be a validation error

As a follow up we could get rid of any RFC validations we're doing with the egulias package as it's the same validator used by Symfony in the `Address::__construct`

<details>
<summary>Done</summary>
<br />

- [x] lib/private/Mail/Message.php:    * SwiftMailer does currently not work with IDN domains, this function therefore converts the domains
- [x] lib/private/Mail/Message.php:    * FIXME: Remove this once SwiftMailer supports IDN
- [x] lib/private/Mail/Mailer.php:     * SwiftMailer does currently not work with IDN domains, this function therefore converts the domains
- [x] lib/private/Mail/Mailer.php:     * FIXME: Remove this once SwiftMailer supports IDN
- [x] lib/private/Mail/Mailer.php:            return new Attachment(new \Swift_Attachment($data, $filename, $contentType));
- [x] lib/private/Mail/Mailer.php:            return new Attachment(\Swift_Attachment::fromPath($path, $contentType));
- [x] lib/private/Mail/Mailer.php:                    $mailLogger = new \Swift_Plugins_Loggers_ArrayLogger();
- [x] lib/private/Mail/Mailer.php:                    $mailer->registerPlugin(new \Swift_Plugins_LoggerPlugin($mailLogger));
- [x] lib/private/Mail/Mailer.php:            $mailer->send($message->getSwiftMessage(), $failedRecipients);
- [x] lib/private/Mail/Message.php: * Class Message provides a wrapper around SwiftMail
- [x] lib/private/Mail/Message.php:           $this->swiftMessage->attach($attachment->getSwiftAttachment());
- [x] lib/private/Mail/Message.php:    * Get's the underlying SwiftMessage
- [x] lib/private/Mail/Message.php:   public function setSwiftMessage(\Symfony\Component\Mime\Email $swiftMessage): void {
- [x] lib/private/Mail/Message.php:    * Get's the underlying SwiftMessage
- [x] lib/private/Mail/Message.php:   public function getSwiftMessage(): \Symfony\Component\Mime\Email {
- [x] lib/private/Mail/Attachment.php:        /** @var \Swift_Mime_Attachment */
- [x] lib/private/Mail/Attachment.php:        public function __construct(\Swift_Mime_Attachment $attachment) {
- [x] lib/private/Mail/Attachment.php:         * @return \Swift_Mime_Attachment
- [x] lib/private/Mail/Attachment.php:        public function getSwiftAttachment(): \Swift_Mime_Attachment {

</details>

<details>
<summary>Nothing to do with SwiftMailer (== Done)</summary>
<br />

- [x] lib/private/SystemConfig.php:                           // Legacy Swift (https://github.com/nextcloud/server/pull/17696#discussion_r341302207)
- [x] lib/private/SystemConfig.php:                           // Swift v2
- [x] lib/private/SystemConfig.php:                           // Swift v3
- [x] lib/private/SystemConfig.php:                           // Swift v2
- [x] lib/private/SystemConfig.php:                           // Swift v3

</details>

### CODE REVIEW

**Comment by come-nc** [comment]
Thir rector is clearly not enough for the migration, left instances of Swift in the code:
```
lib/private/Mail/Mailer.php:            return new Attachment(new \Swift_Attachment($data, $filename, $contentType));
lib/private/Mail/Mailer.php:            return new Attachment(\Swift_Attachment::fromPath($path, $contentType));
lib/private/Mail/Mailer.php:                    $mailLogger = new \Swift_Plugins_Loggers_ArrayLogger();
lib/private/Mail/Mailer.php:                    $mailer->registerPlugin(new \Swift_Plugins_LoggerPlugin($mailLogger));
lib/private/Mail/Mailer.php:            $mailer->send($message->getSwiftMessage(), $failedRecipients);
lib/private/Mail/Mailer.php:     * SwiftMailer does currently not work with IDN domains, this function therefore converts the domains
lib/private/Mail/Mailer.php:     * FIXME: Remove this once SwiftMailer supports IDN
lib/private/Mail/Message.php: * Class Message provides a wrapper around SwiftMail
lib/private/Mail/Message.php:           $this->swiftMessage->attach($attachment->getSwiftAttachment());
lib/private/Mail/Message.php:    * SwiftMailer does currently not work with IDN domains, this function therefore converts the domains
lib/private/Mail/Message.php:    * FIXME: Remove this once SwiftMailer supports IDN
lib/private/Mail/Message.php:    * Get's the underlying SwiftMessage
lib/private/Mail/Message.php:   public function setSwiftMessage(\Symfony\Component\Mime\Email $swiftMessage): void {
lib/private/Mail/Message.php:    * Get's the underlying SwiftMessage
lib/private/Mail/Message.php:   public function getSwiftMessage(): \Symfony\Component\Mime\Email {
lib/private/Mail/Attachment.php:        /** @var \Swift_Mime_Attachment */
lib/private/Mail/Attachment.php:        public function __construct(\Swift_Mime_Attachment $attachment) {
lib/private/Mail/Attachment.php:         * @return \Swift_Mime_Attachment
lib/private/Mail/Attachment.php:        public function getSwiftAttachment(): \Swift_Mime_Attachment {
lib/private/SystemConfig.php:                           // Legacy Swift (https://github.com/nextcloud/server/pull/17696#discussion_r341302207)
lib/private/SystemConfig.php:                           // Swift v2
lib/private/SystemConfig.php:                           // Swift v3
lib/private/SystemConfig.php:                           // Swift v2
lib/private/SystemConfig.php:                           // Swift v3
```

**Comment by come-nc** [comment]
Known missing things:

- [ ] setStreamOptions support. This seems not possible with symfony/mailer. https://github.com/nextcloud/server/issues/12702 seems to suggest that it is mainly used to accept self-signed certificate, maybe we can look into supporting that one usecase if possible.
- [ ] timeout setting is currently commented out. Probably possible to fix that but I did not find yet where.

**Review by ChristophWurst — APPROVED** [review]

**Comment by come-nc** [comment]
Stream options and timeout support are back. Just have to use getStream on the transport and work directly on the stream.

So, tests are green, but I did not actually test that this is able to send any email. Anyone has a setup to test this? Does the CI test it?

**Comment by come-nc** [comment]
Ouh, there is still something fishy, the send method is supposed to return the failedRecipient but I did not find how to do that.

**Comment by come-nc** [comment]
> Ouh, there is still something fishy, the send method is supposed to return the failedRecipient but I did not find how to do that.

Also lots of problems spotted by psalm.

- [ ] `mail_smtpauthtype` option config will have to be deprecated, symfony/mailer select authmethod automatically and do not support forcing it.

**Comment by come-nc** [comment]
From the code:
```
	 * SwiftMailer does currently not work with IDN domains, this function therefore converts the domains
	 * FIXME: Remove this once SwiftMailer supports IDN
```
- [ ] Check if symfony/mailer support IDN and we can get rid of this

**Comment by come-nc** [comment]
> Ouh, there is still something fishy, the send method is supposed to return the failedRecipient but I did not find how to do that.

There is no way to do that so it returns an empty array. The method description said this is not always available anyway, and errors can be thrown when sending fails.

**Review by miaulalala — APPROVED** [review]

File: lib/private/Mail/Mailer.php [hunk 1]
```diff
@@ -255,31 +254,39 @@ protected function getInstance(): \Swift_Mailer {
...
-	protected function getSmtpInstance(): \Swift_SmtpTransport {
-		$transport = new \Swift_SmtpTransport();
-		$transport->setTimeout($this->config->getSystemValue('mail_smtptimeout', 10));
-		$transport->setHost($this->config->getSystemValue('mail_smtphost', '127.0.0.1'));
-		$transport->setPort($this->config->getSystemValue('mail_smtpport', 25));
+	protected function getSmtpInstance(): EsmtpTransport {
+		$transport = new EsmtpTransport(
+			$this->config->getSystemValue('mail_smtphost', '127.0.0.1'),
+			$this->config->getSystemValue('mail_smtpport', 25),
+			null,
```
Comment (come-nc) [inline lib/private/Mail/Mailer.php:269]: This null is a boolean used to indicate if tls should be used.
If null is given, the logic is:
```php
        if (null === $tls) {
            if (465 === $port) {
                $tls = true;
            } else {
                $tls = \defined('OPENSSL_VERSION_NUMBER') && 0 === $port && 'localhost' !== $host;
            }
        }
        if (!$tls) {
            $stream->disableTls();
        }
        if (0 === $port) {
            $port = $tls ? 465 : 25;
        }
```

It’s not clear for me if we should keep passing null or if we should somehow plug `mail_smtpsecure` config option in there.

**Comment by come-nc** [comment]
TODO:
- Figure out if `mail_smtpsecure` should be dropped or re-implemented (see https://github.com/nextcloud/server/pull/30349#issuecomment-1268107105 and https://github.com/nextcloud/server/pull/30349#discussion_r1028074804 )
- Drop `mail_smtpauthtype`
- Test that mail sending still works
- Test as much as possible
- Remove workaround code pointed by https://github.com/nextcloud/server/pull/30349#issuecomment-1268135892 and see if it still works for addresses with IDN in it. If yes remove this workaround. If not update the comment.
- Document changes in behavior, API and configuration.

@PVince81 @ChristophWurst I’d say maybe a week of work?

**Comment by miaulalala** [comment]
We might be able to remove the egulias package as symphony mailer includes its own validator: https://github.com/symfony/symfony/blob/6.2/src/Symfony/Component/Validator/Constraints/EmailValidator.php

File: lib/private/Mail/Mailer.php (hunk 1 above)
Comment (miaulalala) [reply lib/private/Mail/Mailer.php:269]: STARTTLS with port 578 can't be enforced any longer. The automatic config will use it if the server advertises STARTTLS, but there is no config option any longer.

If explicit TLS (465) in the config.php or settings is set, we still use that and set the $tls param to true.

File: lib/private/Mail/Message.php [hunk 2]
```diff
@@ -264,6 +260,40 @@ public function setBody(string $body, string $contentType): IMessage {
...
+		$this->symfonyEmail->to(...$this->convertAddresses($this->getTo()));
+		$this->symfonyEmail->from(...$this->convertAddresses($this->getFrom()));
+
+		if (!empty($this->getReplyTo())) {
+			$this->symfonyEmail->replyTo(...$this->convertAddresses($this->getReplyTo()));
+		}
+		if (!empty($this->getCc())) {
+			$this->symfonyEmail->cc(...$this->convertAddresses($this->getCc()));
+		}
+		if (!empty($this->getBcc())) {
```
Comment (ChristophWurst) [inline lib/private/Mail/Message.php:292]: check can probably be skipped if passing empty arrays into cc/bcc is allowed

**Comment by ChristophWurst** [comment]
Tested with my personal SMTP test account that had already been set up for older releases and it still works :+1:

**Comment by juliushaertl** [comment]
https://github.com/nextcloud/3rdparty/pull/1179 is merged

File: lib/private/Mail/Mailer.php [hunk 3]
```diff
@@ -162,11 +169,11 @@ public function createEMailTemplate(string $emailId, array $data = []): IEMailTe
...
-	 * @param IMessage|Message $message Message to send
-	 * @return string[] Array with failed recipients. Be aware that this depends on the used mail backend and
-	 * therefore should be considered
-	 * @throws \Exception In case it was not possible to send the message. (for example if an invalid mail address
-	 * has been supplied.)
+	 * If sending failed, the recipients that failed will be returned (to, cc and bcc).
+	 * Will output additional debug info if 'mail_smtpdebug' => 'true' is set in config.php
+	 *
+	 * @param IMessage $message Message to send
+	 * @return array []|string[] $failedRecipients
```
Comment (come-nc) [inline lib/private/Mail/Mailer.php:176]:
```suggestion
	 * @return string[] $failedRecipients
```

**Comment by come-nc** [comment]
(I cannot approve as author of the PR, but I hereby approve what @miaulalala did :+1: )

File: lib/private/Mail/Mailer.php (hunk 3 above)
Comment (miaulalala) [reply lib/private/Mail/Mailer.php:176]: but it can also be an empty array when sending works

Comment (tcitworld) [reply lib/private/Mail/Mailer.php:176]: `string[]` covers empty arrays already :)

**Comment by miaulalala** [comment]
documentation ticket here: https://github.com/nextcloud/documentation/issues/9600

File: lib/private/Mail/Mailer.php (hunk 3 above)
Comment (miaulalala) [reply lib/private/Mail/Mailer.php:176]: I didn't know that! will change it

**Comment by ChristophWurst** [comment]
> There were 10 errors:

File: apps/settings/lib/Controller/MailSettingsController.php [hunk 4]
```diff
@@ -36,6 +36,7 @@
 use OCP\IURLGenerator;
 use OCP\IUserSession;
 use OCP\Mail\IMailer;
+use Symfony\Component\Mailer\Exception\TransportExceptionInterface;
```
Comment (ChristophWurst) [inline apps/settings/lib/Controller/MailSettingsController.php:39]: unused?

File: config/config.sample.php [hunk 5]
```diff
@@ -479,12 +482,9 @@
...
 
 /**
- * This depends on ``mail_smtpmode``. If SMTP authentication is required, choose
- * the authentication type as ``LOGIN`` or ``PLAIN``.
- *
- * Defaults to ``LOGIN``
+ * @deprecated 26.0.0  No longer used
  */
-'mail_smtpauthtype' => 'LOGIN',
+// 'mail_smtpauthtype' => 'LOGIN',
```
Comment (ChristophWurst) [inline config/config.sample.php:487]: drop if not used anymore?

File: lib/public/Mail/IMailer.php [hunk 6]
```diff
@@ -96,7 +96,8 @@ public function createEMailTemplate(string $emailId, array $data = []): IEMailTe
 	public function send(IMessage $message): array;
 
 	/**
-	 * Checks if an e-mail address is valid
+	 * @deprecated 26.0.0 Implicit validation is done in \OC\Mail\Message::setRecipients
```
Comment (ChristophWurst) [inline lib/public/Mail/IMailer.php:99]: just deprecate. the rest is internal info no API user should depend on

File: lib/public/Mail/IMessage.php [hunk 7]
```diff
@@ -50,6 +50,14 @@ public function attach(IAttachment $attachment): IMessage;
 	 */
 	public function setFrom(array $addresses): IMessage;
 
+
+	/**
+	 * Get the from address of this message.
+	 * @return array [] | string[]
```
Comment (ChristophWurst) [inline lib/public/Mail/IMessage.php:56]: `array[]` means two-dimensional array. It is the same as `mixed[][]`.

do we need the getter in the public interface?

File: lib/public/Mail/IMessage.php [hunk 8]
```diff
@@ -86,6 +108,27 @@ public function setCc(array $recipients): IMessage;
...
+	 * @since 26.0.0
+	 */
+	public function setSubject(string $subject): IMessage;
+
+	/**
+	 * Get the "from" subject of this message.
+	 * @return string
+	 * @since 26.0.0
+	 */
+	public function getSubject(): string;
```
Comment (ChristophWurst) [inline lib/public/Mail/IMessage.php:130]: only expose if really required on the public interface

**Review by ChristophWurst — COMMENTED** [review]
Looks sane otherwise

File: lib/public/Mail/IMessage.php (hunk 8 above)
Comment (ChristophWurst) [reply lib/public/Mail/IMessage.php:130]: elaborate

Comment (ChristophWurst) [reply lib/public/Mail/IMessage.php:130]: In `send` you get an IMessage. Throw an InvalidArgumentException if someone passes an object that is not an instance of the Message implementation created by `createMessage`. After this check you can safely assume that you have a Message, not just an IMessage and call Message::getSubject

File: lib/public/Mail/IMailer.php (hunk 6 above)
Comment (tcitworld) [reply lib/public/Mail/IMailer.php:99]: I don't think it should be depreciated though. There are valid uses of checking if an email's syntax is valid without sending an email at the same time.

File: lib/public/Mail/IMessage.php (hunk 8 above)
Comment (tcitworld) [reply lib/public/Mail/IMessage.php:130]: Could be useful in some kind of treatment to the emails after they're built, such as prefixing them.

Comment (tcitworld) [reply lib/public/Mail/IMessage.php:130]: It's not *needed* though, it's fine

Comment (ChristophWurst) [reply lib/public/Mail/IMessage.php:130]: Let's expose the getter when we need it. Until then: YAGNI :)

File: lib/public/Mail/IMailer.php (hunk 6 above)
Comment (ChristophWurst) [reply lib/public/Mail/IMailer.php]: You are right. This migration has nothing to do with the validation. Validation can still make sense without sending a message :+1:

File: lib/public/Mail/IMailer.php [hunk 9]
```diff
@@ -96,7 +96,7 @@ public function createEMailTemplate(string $emailId, array $data = []): IEMailTe
 	public function send(IMessage $message): array;
 
 	/**
-	 * Checks if an e-mail address is valid
+	 * @deprecated 26.0.0
```
Comment (ChristophWurst) [inline lib/public/Mail/IMailer.php:99]: let's undo this. email validation is still useful

**Review by ChristophWurst — CHANGES_REQUESTED** [review]

**Review by ChristophWurst — APPROVED** [review]

**Comment by miaulalala** [comment]
@DaphneMuller https://github.com/nextcloud/documentation/issues/9600 is the documentation ticket

**Comment by joshtrichards** [comment]
Indirectly completes #29333

**Pull request merged**